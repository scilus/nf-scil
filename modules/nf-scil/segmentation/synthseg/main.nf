process SEGMENTATION_SYNTHSEG {
    tag "$meta.id"
    label 'process_single'

    container "freesurfer/freesurfer:7.4.1"
    containerOptions "--entrypoint ''"

    input:
    tuple val(meta), path(image), path(lesion)

    output:
    tuple val(meta), path("*.bam"), emit: bam
    // TODO nf-core: List additional required output channels/values here
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    def gpu = task.ext.gpu ? "" : "--cpu"
    def parc = task.ext.parc ? "--parc" : ""
    def robust = task.ext.robust ? "--robust" : ""
    def fast = task.ext.fast ? "--fast" : ""
    def ct = task.ext.ct ? "--ct" : ""
    //TODO add all the others parameter with the corresponding optional outputs

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    mri_synthseg --i $image --o t1.nii.gz $--threads 1 gpu $parc $robust $fast $ct

    # WM Mask
    mri_binarize --i t1.nii.gz \
                --match 2 7 12 16 28 41 46 49 51 60 \
                --o ${prefix}__mask_wm.ni.gz

    # GM Mask
    mri_binarize --i t1.nii.gz \
                --match 3 8 11 17 18 26 42 47 50 \
                --o ${prefix}__mask_wm.ni.gz

    # CSF Mask
    mri_binarize --i t1.nii.gz \
                --match 4 5 14 15 24 43 44 \
                --o ${prefix}__mask_wm.ni.gz

    if [[ -f "$lesion" ]];
    then
        mri_binarize --i ${prefix}__mask_wm.ni.gz --union $lesion --o ${prefix}__mask_wm.ni.gz
    fi

    mri_convert --i ${prefix}__mask_wm.ni.gz --out_data_type uchar --o ${prefix}__mask_wm.ni.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        segmentation: \$(samtools --version |& sed '1!d ; s/samtools //')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    // TODO nf-core: A stub section should mimic the execution of the original module as best as possible
    //               Have a look at the following examples:
    //               Simple example: https://github.com/nf-core/modules/blob/818474a292b4860ae8ff88e149fbcda68814114d/modules/nf-core/bcftools/annotate/main.nf#L47-L63
    //               Complex example: https://github.com/nf-core/modules/blob/818474a292b4860ae8ff88e149fbcda68814114d/modules/nf-core/bedtools/split/main.nf#L38-L54
    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        segmentation: \$(samtools --version |& sed '1!d ; s/samtools //')
    END_VERSIONS
    """
}
