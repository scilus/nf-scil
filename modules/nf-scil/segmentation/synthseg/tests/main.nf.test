nextflow_process {

    name "Test Process SEGMENTATION_SYNTHSEG"
    script "../main.nf"
    process "SEGMENTATION_SYNTHSEG"

    tag "modules"
    tag "modules_nfcore"
    tag "segmentation"
    tag "segmentation/synthseg"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "freesurfer.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("segmentation - synthseg - basic") {
        config "./nextflow_basic.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/t1.nii.gz", checkIfExists: true,),
                        [],
                        file("\${test_data_directory}/license.txt", checkIfExists: true,)
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }

            )
        }
    }

    test("segmentation - synthseg - options") {
        config "./nextflow_options.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/t1.nii.gz", checkIfExists: true,),
                        [],
                        file("\${test_data_directory}/license.txt", checkIfExists: true,)
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }

            )
        }
    }

    // test("segmentation - synthseg - lesion") {
    //     config "./nextflow_lesion.config"
    //     when {
    //         process {
    //             """
    //             input[0] = LOAD_DATA.out.test_data_directory.map{
    //                 test_data_directory -> [
    //                     [ id:'test', single_end:false ], // meta map
    //                     file("\${test_data_directory}/t1.nii.gz", checkIfExists: true,),
    //                     file("\${test_data_directory}/lesion.nii.gz", checkIfExists: true,),
    //                     file("\${test_data_directory}/license.txt", checkIfExists: true,)

    //                 ]
    //             }
    //             """
    //         }
    //     }

    //     then {
    //         assertAll(
    //             { assert process.success },
    //             { assert snapshot(process.out).match() }
    //         )
    //     }
    // }

}
