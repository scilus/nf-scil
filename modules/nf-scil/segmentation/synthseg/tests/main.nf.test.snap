{
    "segmentation - synthseg": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,5aa95919bfa5b1dd2f8f7a0ae2667bfb"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,432d6a28a10823c0ca5b2f40d2353e2f"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,e60ff1b5847502a1e3f8c4a590d4a73c"
                    ]
                ],
                "3": [
                    "versions.yml:md5,b65418e9ce2ab1cce590b1f659e18c1d"
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,e60ff1b5847502a1e3f8c4a590d4a73c"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,432d6a28a10823c0ca5b2f40d2353e2f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b65418e9ce2ab1cce590b1f659e18c1d"
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,5aa95919bfa5b1dd2f8f7a0ae2667bfb"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-06T19:27:40.416183"
    }
}